django/apps/registry.py

```diff
@@ -44,7 +44,8 @@ def __init__(self, installed_apps=()):
         self.apps_ready = self.models_ready = self.ready = False
 
         # Lock for thread-safe population.
-        self._lock = threading.Lock()
+        self._lock = threading.RLock()
+        self._populate_running = False
```

Comment:
It's a shame that `RLock` doesn't expose its `_count` attribute (in the Python implementation; I didn't check the C implementation but it certainly has a variable that fulfills the same role).

A comment stating that's the reason why you need a separate `_populate_running` to track whether the lock is acquired for the first time or again by the same thread would help.